cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -846,19 +842,32 @@ private RunRecordMeta recordProgramStop(ProgramRunId programRunId, long stopTs,
    * once it is already running.
    *
    * @param existing the existing run record meta of the given program run
+   * @param nextProgramState the program state to transition to
+   * @param nextClusterState the cluster state to transition to
    * @param sourceId unique id representing the source of program run status, such as the message id of the program
    *                 run status notification in TMS. The source id must increase as the recording time of the program
    *                 run status increases, so that the attempt to persist program run status older than the existing
    *                 program run status will be ignored
-   * @param recordType the type of record corresponding to the current status
    * @return {@code true} if the program run is allowed to persist the given status, {@code false} otherwise
    */
-  private boolean isValid(RunRecordMeta existing, byte[] sourceId, String recordType) {
+  private boolean isValid(@Nullable RunRecordMeta existing, ProgramRunStatus nextProgramState,
+                          ProgramRunClusterStatus nextClusterState, byte[] sourceId) {
     byte[] existingSourceId = existing.getSourceId();
```

Comment:
This would have a NPE if `existing` is null (which it is `@Nullable` now).

Reply:
this is incorrectly annotated. Was trying to consolidate some logic but decided not to and forgot to remove. Will remove annotation.

Reply:
Has this removed?